### ISSUE TITLE
Jetty key-store hot reload fails

### ISSUE THREAD

**Issue by Maurice-Betzel** [ops4j/org.ops4j.pax.web#1811]
Hi all,

does pax-web support the Jetty key store hot reload mechanism, what I am seeing is that after a reload the connector is not restarted.
The restore also does not run on a pax web thread:

```
2022-12-20T17:03:15,751 | DEBUG | paxweb-config-3-thread-3 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Selected Protocols [TLSv1.2] of [TLSv1.3, TLSv1.2, TLSv1.1, TLSv1, SSLv3, SSLv2Hello]
2022-12-20T17:03:15,752 | DEBUG | paxweb-config-3-thread-3 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Selected Ciphers   [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256, TLS_ECDH_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDH_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256, TLS_
2022-12-20T17:03:15,752 | DEBUG | paxweb-config-3-thread-3 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Customize sun.security.ssl.SSLEngineImpl@569ce10c
2022-12-20T17:03:15,753 | DEBUG | paxweb-config-3-thread-3 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Customize sun.security.ssl.SSLEngineImpl@2df0016a
2022-12-20T17:03:15,769 | INFO  | paxweb-config-3-thread-3 | AbstractConnector                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Started 0.0.0.0:8444@4b3be26c{SSL, (ssl, alpn, acme-tls/1, h2, http/1.1)}{0.0.0.0:8444}
2022-12-20T17:03:15,770 | INFO  | paxweb-config-3-thread-3 | Server                           | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Started @435123ms

After update and reload no Started state and the ACME challenge failed:

2022-12-20T17:03:24,817 | INFO  | AcmeQuartzScheduler_Worker-1 | ACMERenewJob                     | 195 - eu.abeel.platform.security.acme.api - 1.0.0 | Authorization for pending domain localhost
2022-12-20T17:03:30,689 | DEBUG | Scanner-2        | KeyStoreScanner                  | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | changed C:\Users\betzm\Development\karaf\runtime-3.2.1\etc\keystore\digital.gaston-schul.com.jks
2022-12-20T17:03:30,690 | DEBUG | Scanner-2        | KeyStoreScanner                  | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | reloading keystore file C:\Users\betzm\Development\karaf\runtime-3.2.1\etc\keystore\digital.gaston-schul.com.jks
2022-12-20T17:03:30,697 | DEBUG | Scanner-2        | X509                             | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Certificate alias=digital.gaston-schul.com SAN dns=digital.gaston-schul.com in X509@7ca019cc(digital.gaston-schul.com,h=[],a=[],w=[])
2022-12-20T17:03:30,698 | DEBUG | Scanner-2        | X509                             | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Certificate CN alias=digital.gaston-schul.com CN=digital.gaston-schul.com in X509@7ca019cc(digital.gaston-schul.com,h=[digital.gaston-schul.com],a=[],w=[])
2022-12-20T17:03:30,699 | INFO  | Scanner-2        | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | x509=X509@7ca019cc(digital.gaston-schul.com,h=[digital.gaston-schul.com],a=[],w=[]) for Server@10373d79[provider=null,keyStore=file:///C:/Users/betzm/Development/karaf/runtime-3.2.1/etc/keystore/digital.gaston-schul.com.jks,trustStore=null]
2022-12-20T17:03:30,700 | DEBUG | Scanner-2        | X509                             | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Certificate alias=digital.gaston-schul.com (r3) SAN dns=digital.gaston-schul.com in X509@7267e828(digital.gaston-schul.com (r3),h=[],a=[],w=[])
2022-12-20T17:03:30,701 | DEBUG | Scanner-2        | X509                             | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Certificate CN alias=digital.gaston-schul.com (r3) CN=digital.gaston-schul.com in X509@7267e828(digital.gaston-schul.com (r3),h=[digital.gaston-schul.com],a=[],w=[])
2022-12-20T17:03:30,701 | INFO  | Scanner-2        | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | x509=X509@7267e828(digital.gaston-schul.com (r3),h=[digital.gaston-schul.com],a=[],w=[]) for Server@10373d79[provider=null,keyStore=file:///C:/Users/betzm/Development/karaf/runtime-3.2.1/etc/keystore/digital.gaston-schul.com.jks,trustStore=null]
2022-12-20T17:03:30,702 | DEBUG | Scanner-2        | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | managers=[org.eclipse.jetty.util.ssl.SniX509ExtendedKeyManager@335ae213] for Server@10373d79[provider=null,keyStore=file:///C:/Users/betzm/Development/karaf/runtime-3.2.1/etc/keystore/digital.gaston-schul.com.jks,trustStore=null]
2022-12-20T17:03:30,704 | DEBUG | Scanner-2        | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Selected Protocols [TLSv1.2] of [TLSv1.3, TLSv1.2, TLSv1.1, TLSv1, SSLv3, SSLv2Hello]
2022-12-20T17:03:30,704 | DEBUG | Scanner-2        | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Selected Ciphers   [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256, TLS_ECDH_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDH_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDH_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDH_RSA_WITH_AES_128_GCM_SHA256, TLS_
2022-12-20T17:03:40,981 | DEBUG | qtp1523161524-231 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Customize sun.security.ssl.SSLEngineImpl@23ef02c9
2022-12-20T17:03:40,989 | DEBUG | qtp1523161524-233 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | SNI matching for type=host_name (0), value=localhost
2022-12-20T17:03:40,990 | DEBUG | qtp1523161524-233 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | SNI host name localhost
2022-12-20T17:03:40,992 | DEBUG | qtp1523161524-233 | SniX509ExtendedKeyManager        | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Chose explicit alias=null keyType=EC on sun.security.ssl.SSLEngineImpl@23ef02c9
2022-12-20T17:03:40,993 | DEBUG | qtp1523161524-233 | SniX509ExtendedKeyManager        | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Chose explicit alias=null keyType=EC on sun.security.ssl.SSLEngineImpl@23ef02c9
2022-12-20T17:03:40,993 | DEBUG | qtp1523161524-233 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Selecting alias: keyType=RSA, sni=localhost, sniRequired=false, certs=[X509@7267e828(digital.gaston-schul.com (r3),h=[digital.gaston-schul.com],a=[],w=[])]
2022-12-20T17:03:40,995 | DEBUG | qtp1523161524-233 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Selected alias=delegate_no_sni_match
2022-12-20T17:03:40,995 | DEBUG | qtp1523161524-233 | SniX509ExtendedKeyManager        | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Chose delegate alias=digital.gaston-schul.com (r3) keyType=RSA on sun.security.ssl.SSLEngineImpl@23ef02c9
2022-12-20T17:03:42,972 | DEBUG | qtp1523161524-233 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Customize sun.security.ssl.SSLEngineImpl@2773d525
2022-12-20T17:03:42,982 | DEBUG | qtp1523161524-217 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | SNI matching for type=host_name (0), value=localhost
2022-12-20T17:03:42,983 | DEBUG | qtp1523161524-217 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | SNI host name localhost
2022-12-20T17:03:42,985 | DEBUG | qtp1523161524-217 | SniX509ExtendedKeyManager        | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Chose explicit alias=null keyType=EC on sun.security.ssl.SSLEngineImpl@2773d525
2022-12-20T17:03:42,986 | DEBUG | qtp1523161524-217 | SniX509ExtendedKeyManager        | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Chose explicit alias=null keyType=EC on sun.security.ssl.SSLEngineImpl@2773d525
2022-12-20T17:03:42,987 | DEBUG | qtp1523161524-217 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Selecting alias: keyType=RSA, sni=localhost, sniRequired=false, certs=[X509@7267e828(digital.gaston-schul.com (r3),h=[digital.gaston-schul.com],a=[],w=[])]
2022-12-20T17:03:42,988 | DEBUG | qtp1523161524-217 | SslContextFactory                | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Selected alias=delegate_no_sni_match
2022-12-20T17:03:42,990 | DEBUG | qtp1523161524-217 | SniX509ExtendedKeyManager        | 100 - org.eclipse.jetty.util - 9.4.43.v20210629 | Chose delegate alias=digital.gaston-schul.com (r3) keyType=RSA on sun.security.ssl.SSLEngineImpl@2773d525
2022-12-20T17:03:43,002 | INFO  | AcmeQuartzScheduler_Worker-1 | JobRunShell                      | 203 - org.quartz-scheduler.quartz - 2.3.2 | Job acme.action.job.group.ACMERenewJob threw a JobExecutionException: 
org.quartz.JobExecutionException: Challenge has failed due to Failed to connect to [::1]:8444 for the tls-alpn-01 challenge
	at eu.abeel.platform.security.acme.api.job.ACMERenewJob.executeJob(ACMERenewJob.java:126) ~[?:?]
	at eu.abeel.platform.security.acme.api.spi.ACMEJob.execute(ACMEJob.java:42) ~[?:?]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202) [!/:?]
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573) [!/:?]
2022-12-20T17:03:43,004 | WARN  | AcmeQuartzScheduler_Worker-1 | ACMEActionJobListener            | 195 - eu.abeel.platform.security.acme.api - 1.0.0 | Job ACMERenewJob in group acme.action.job.group executed in 18336 ms
2022-12-20T17:03:43,004 | WARN  | AcmeQuartzScheduler_Worker-1 | ACMEActionJobListener            | 195 - eu.abeel.platform.security.acme.api - 1.0.0 | Challenge has failed due to Failed to connect to [::1]:8444 for the tls-alpn-01 challenge
```

**Comment by grgrzybek** [comment]
Hmm, indeed, `Scanner-2` seems to be a Jetty thread, which I'm not aware of. Jetty seems to detect changed keystore, but Pax Web doesn't notice it.
Also your thread `AcmeQuartzScheduler_Worker-1` seems to hit Jetty in the middle of some process.

Pax Web wasn't designed to handle such restart (it'd have to work consistently for Jetty, Tomcat and Undertow).

For now I leave this issue open and when I'm back from holidays, I'll check what can be done.

Thanks for report!

**Comment by Maurice-Betzel** [comment]
Hi Grzegorz,
the ACMEQuartzScheduler is my Quartz job initiating the ACME TLS validation and storing the challenge key-pair I received from the ACME provider, then waits for a short period for Jetty to pick up the updated key-store, and then trigger the challenge on the ACME provider which then executes a acme-tls ssl handshake where Jetty should return the previously stored certificate of the key-pair.

**Comment by Maurice-Betzel** [comment]
I guess a quick fix would be to restart the Pax Web Jetty bundle after a key-store update, avoiding the Jetty internal hot reload of the key-store?

**Comment by grgrzybek** [comment]
I'll handle it somehow - but in January, ok?

**Comment by Maurice-Betzel** [comment]
You are on holidays, so go and have fun. I am talking to myself presenting the latest findings to the world and having to apologize to you, as recent testing confirmed that pax web does not have to do anything for Jetty on a hot reload of the key-store.
I found out that the ACME challenge was not finished due to an SSL handshake error because I corrupted the key-store with my code. So I tested it with installing a simple Servlet and manipulating the key-store by hand swapping out certificates, with hot reload activated, and the browser responded each time with the certificate that was stored in the key-store.
Conclusion for Pax Web Jetty: hot reloading of key-stores seems not to be a problem.
